// win32-def/dist/lib/consts/winspool/PRINTER_STATUS.enum.js
/**
 * @link https://learn.microsoft.com/zh-cn/windows/win32/printdocs/printer-info-6
 * @link https://github.com/wine-mirror/wine/blob/master/include/winspool.h
 */
export var PRINTER_STATUS;
(function (PRINTER_STATUS) {
    PRINTER_STATUS[PRINTER_STATUS["unknown"] = 0] = "unknown";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PAUSED"] = 1] = "PRINTER_STATUS_PAUSED";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_ERROR"] = 2] = "PRINTER_STATUS_ERROR";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PENDING_DELETION"] = 4] = "PRINTER_STATUS_PENDING_DELETION";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PAPER_JAM"] = 8] = "PRINTER_STATUS_PAPER_JAM";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PAPER_OUT"] = 16] = "PRINTER_STATUS_PAPER_OUT";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_MANUAL_FEED"] = 32] = "PRINTER_STATUS_MANUAL_FEED";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PAPER_PROBLEM"] = 64] = "PRINTER_STATUS_PAPER_PROBLEM";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_OFFLINE"] = 128] = "PRINTER_STATUS_OFFLINE";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_IO_ACTIVE"] = 256] = "PRINTER_STATUS_IO_ACTIVE";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_BUSY"] = 512] = "PRINTER_STATUS_BUSY";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PRINTING"] = 1024] = "PRINTER_STATUS_PRINTING";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_OUTPUT_BIN_FULL"] = 2048] = "PRINTER_STATUS_OUTPUT_BIN_FULL";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_NOT_AVAILABLE"] = 4096] = "PRINTER_STATUS_NOT_AVAILABLE";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_WAITING"] = 8192] = "PRINTER_STATUS_WAITING";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PROCESSING"] = 16384] = "PRINTER_STATUS_PROCESSING";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_INITIALIZING"] = 32768] = "PRINTER_STATUS_INITIALIZING";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_WARMING_UP"] = 65536] = "PRINTER_STATUS_WARMING_UP";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_TONER_LOW"] = 131072] = "PRINTER_STATUS_TONER_LOW";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_NO_TONER"] = 262144] = "PRINTER_STATUS_NO_TONER";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_PAGE_PUNT"] = 524288] = "PRINTER_STATUS_PAGE_PUNT";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_USER_INTERVENTION"] = 1048576] = "PRINTER_STATUS_USER_INTERVENTION";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_OUT_OF_MEMORY"] = 2097152] = "PRINTER_STATUS_OUT_OF_MEMORY";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_DOOR_OPEN"] = 4194304] = "PRINTER_STATUS_DOOR_OPEN";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_SERVER_UNKNOWN"] = 8388608] = "PRINTER_STATUS_SERVER_UNKNOWN";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_POWER_SAVE"] = 16777216] = "PRINTER_STATUS_POWER_SAVE";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_SERVER_OFFLINE"] = 33554432] = "PRINTER_STATUS_SERVER_OFFLINE";
    PRINTER_STATUS[PRINTER_STATUS["PRINTER_STATUS_DRIVER_UPDATE_NEEDED"] = 67108864] = "PRINTER_STATUS_DRIVER_UPDATE_NEEDED";
})(PRINTER_STATUS || (PRINTER_STATUS = {}));
