// win32-def/dist/lib/consts/virtual-key.enum.js
/**
 * Virtual-Key Codes
 *
 * @description The following table shows the symbolic constant names, hexadecimal values,
 * and mouse or keyboard equivalents for the virtual-key codes used by the system.
 * The codes are listed in numeric order.
 * @link https://docs.microsoft.com/zh-cn/windows/win32/inputdev/virtual-key-codes?redirectedfrom=MSDN
 * @link https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes?redirectedfrom=MSDN
 */
export var VirtualKey;
(function (VirtualKey) {
    /** Left mouse button */
    VirtualKey[VirtualKey["VK_LBUTTON"] = 1] = "VK_LBUTTON";
    /** Right mouse button */
    VirtualKey[VirtualKey["VK_RBUTTON"] = 2] = "VK_RBUTTON";
    /** Control-break processing */
    VirtualKey[VirtualKey["VK_CANCEL"] = 3] = "VK_CANCEL";
    /** Middle mouse button(three - button mouse) */
    VirtualKey[VirtualKey["VK_MBUTTON"] = 4] = "VK_MBUTTON";
    /** X1 mouse button */
    VirtualKey[VirtualKey["VK_XBUTTON1"] = 5] = "VK_XBUTTON1";
    /** X2 mouse button */
    VirtualKey[VirtualKey["VK_XBUTTON2"] = 6] = "VK_XBUTTON2";
    /** BACKSPACE key */
    VirtualKey[VirtualKey["VK_BACK"] = 8] = "VK_BACK";
    /** TAB key */
    VirtualKey[VirtualKey["VK_TAB"] = 9] = "VK_TAB";
    /** CLEAR key */
    VirtualKey[VirtualKey["VK_CLEAR"] = 12] = "VK_CLEAR";
    VirtualKey[VirtualKey["VK_RETURN"] = 13] = "VK_RETURN";
    VirtualKey[VirtualKey["VK_SHIFT"] = 16] = "VK_SHIFT";
    VirtualKey[VirtualKey["VK_CONTROL"] = 17] = "VK_CONTROL";
    /** ALT key */
    VirtualKey[VirtualKey["VK_MENU"] = 18] = "VK_MENU";
    VirtualKey[VirtualKey["VK_PAUSE"] = 19] = "VK_PAUSE";
    /** CAPS LOCK key */
    VirtualKey[VirtualKey["VK_CAPITAL"] = 20] = "VK_CAPITAL";
    /** IME Kana mode */
    VirtualKey[VirtualKey["VK_KANA"] = 21] = "VK_KANA";
    /** IME Hanguel mode(maintained for compatibility; use VK_HANGUL)  */
    // VK_HANGUEL = 0x15,
    /** IME Hangul mode */
    VirtualKey[VirtualKey["VK_HANGUL"] = 21] = "VK_HANGUL";
    /** IME On */
    VirtualKey[VirtualKey["VK_IME_ON"] = 22] = "VK_IME_ON";
    VirtualKey[VirtualKey["VK_JUNJA"] = 23] = "VK_JUNJA";
    VirtualKey[VirtualKey["VK_FINAL"] = 24] = "VK_FINAL";
    VirtualKey[VirtualKey["VK_HANJA"] = 25] = "VK_HANJA";
    VirtualKey[VirtualKey["VK_KANJI"] = 25] = "VK_KANJI";
    /** IME Off */
    VirtualKey[VirtualKey["VK_IME_OFF"] = 26] = "VK_IME_OFF";
    /** ESC key */
    VirtualKey[VirtualKey["VK_ESCAPE"] = 27] = "VK_ESCAPE";
    /** IME convert */
    VirtualKey[VirtualKey["VK_CONVERT"] = 28] = "VK_CONVERT";
    /** IME nonconvert */
    VirtualKey[VirtualKey["VK_NONCONVERT"] = 29] = "VK_NONCONVERT";
    /** IME accept */
    VirtualKey[VirtualKey["VK_ACCEPT"] = 30] = "VK_ACCEPT";
    /** IME mode change request */
    VirtualKey[VirtualKey["VK_MODECHANGE"] = 31] = "VK_MODECHANGE";
    /** SPACEBAR */
    VirtualKey[VirtualKey["VK_SPACE"] = 32] = "VK_SPACE";
    /** PAGE UP key */
    VirtualKey[VirtualKey["VK_PRIOR"] = 33] = "VK_PRIOR";
    /** PAGE DOWN key */
    VirtualKey[VirtualKey["VK_NEXT"] = 34] = "VK_NEXT";
    /** END key */
    VirtualKey[VirtualKey["VK_END"] = 35] = "VK_END";
    /** HOME key */
    VirtualKey[VirtualKey["VK_HOME"] = 36] = "VK_HOME";
    /** LEFT ARROW key */
    VirtualKey[VirtualKey["VK_LEFT"] = 37] = "VK_LEFT";
    /** UP ARROW key */
    VirtualKey[VirtualKey["VK_UP"] = 38] = "VK_UP";
    /** RIGHT ARROW key */
    VirtualKey[VirtualKey["VK_RIGHT"] = 39] = "VK_RIGHT";
    /** DOWN ARROW key */
    VirtualKey[VirtualKey["VK_DOWN"] = 40] = "VK_DOWN";
    /** SELECT key */
    VirtualKey[VirtualKey["VK_SELECT"] = 41] = "VK_SELECT";
    /** PRINT key */
    VirtualKey[VirtualKey["VK_PRINT"] = 42] = "VK_PRINT";
    /** EXECUTE key */
    VirtualKey[VirtualKey["VK_EXECUTE"] = 43] = "VK_EXECUTE";
    /** PRINT SCREEN key */
    VirtualKey[VirtualKey["VK_SNAPSHOT"] = 44] = "VK_SNAPSHOT";
    /** INS key */
    VirtualKey[VirtualKey["VK_INSERT"] = 45] = "VK_INSERT";
    /** DEL key */
    VirtualKey[VirtualKey["VK_DELETE"] = 46] = "VK_DELETE";
    /** HELP key */
    VirtualKey[VirtualKey["VK_HELP"] = 47] = "VK_HELP";
    /** 0 key */
    VirtualKey[VirtualKey["VK_0"] = 48] = "VK_0";
    VirtualKey[VirtualKey["VK_1"] = 49] = "VK_1";
    VirtualKey[VirtualKey["VK_2"] = 50] = "VK_2";
    VirtualKey[VirtualKey["VK_3"] = 51] = "VK_3";
    VirtualKey[VirtualKey["VK_4"] = 52] = "VK_4";
    VirtualKey[VirtualKey["VK_5"] = 53] = "VK_5";
    VirtualKey[VirtualKey["VK_6"] = 54] = "VK_6";
    VirtualKey[VirtualKey["VK_7"] = 55] = "VK_7";
    VirtualKey[VirtualKey["VK_8"] = 56] = "VK_8";
    VirtualKey[VirtualKey["VK_9"] = 57] = "VK_9";
    /** A key */
    VirtualKey[VirtualKey["VK_A"] = 65] = "VK_A";
    VirtualKey[VirtualKey["VK_B"] = 66] = "VK_B";
    VirtualKey[VirtualKey["VK_C"] = 67] = "VK_C";
    VirtualKey[VirtualKey["VK_D"] = 68] = "VK_D";
    VirtualKey[VirtualKey["VK_E"] = 69] = "VK_E";
    VirtualKey[VirtualKey["VK_F"] = 70] = "VK_F";
    VirtualKey[VirtualKey["VK_G"] = 71] = "VK_G";
    VirtualKey[VirtualKey["VK_H"] = 72] = "VK_H";
    VirtualKey[VirtualKey["VK_I"] = 73] = "VK_I";
    VirtualKey[VirtualKey["VK_J"] = 74] = "VK_J";
    VirtualKey[VirtualKey["VK_K"] = 75] = "VK_K";
    VirtualKey[VirtualKey["VK_L"] = 76] = "VK_L";
    VirtualKey[VirtualKey["VK_M"] = 77] = "VK_M";
    VirtualKey[VirtualKey["VK_N"] = 78] = "VK_N";
    VirtualKey[VirtualKey["VK_O"] = 79] = "VK_O";
    VirtualKey[VirtualKey["VK_P"] = 80] = "VK_P";
    VirtualKey[VirtualKey["VK_Q"] = 81] = "VK_Q";
    VirtualKey[VirtualKey["VK_R"] = 82] = "VK_R";
    VirtualKey[VirtualKey["VK_S"] = 83] = "VK_S";
    VirtualKey[VirtualKey["VK_T"] = 84] = "VK_T";
    VirtualKey[VirtualKey["VK_U"] = 85] = "VK_U";
    VirtualKey[VirtualKey["VK_V"] = 86] = "VK_V";
    VirtualKey[VirtualKey["VK_W"] = 87] = "VK_W";
    VirtualKey[VirtualKey["VK_X"] = 88] = "VK_X";
    VirtualKey[VirtualKey["VK_Y"] = 89] = "VK_Y";
    VirtualKey[VirtualKey["VK_Z"] = 90] = "VK_Z";
    /** Left Windows key(Natural keyboard) */
    VirtualKey[VirtualKey["VK_LWIN"] = 91] = "VK_LWIN";
    /** Right Windows key(Natural keyboard) */
    VirtualKey[VirtualKey["VK_RWIN"] = 92] = "VK_RWIN";
    /** Applications key(Natural keyboard) */
    VirtualKey[VirtualKey["VK_APPS"] = 93] = "VK_APPS";
    /** Computer Sleep key */
    VirtualKey[VirtualKey["VK_SLEEP"] = 95] = "VK_SLEEP";
    /** Numeric keypad 0 key */
    VirtualKey[VirtualKey["VK_NUMPAD0"] = 96] = "VK_NUMPAD0";
    /** Numeric keypad 1 key */
    VirtualKey[VirtualKey["VK_NUMPAD1"] = 97] = "VK_NUMPAD1";
    /** Numeric keypad 2 key */
    VirtualKey[VirtualKey["VK_NUMPAD2"] = 98] = "VK_NUMPAD2";
    /** Numeric keypad 3 key */
    VirtualKey[VirtualKey["VK_NUMPAD3"] = 99] = "VK_NUMPAD3";
    /** Numeric keypad 4 key */
    VirtualKey[VirtualKey["VK_NUMPAD4"] = 100] = "VK_NUMPAD4";
    /** Numeric keypad 5 key */
    VirtualKey[VirtualKey["VK_NUMPAD5"] = 101] = "VK_NUMPAD5";
    /** Numeric keypad 6 key */
    VirtualKey[VirtualKey["VK_NUMPAD6"] = 102] = "VK_NUMPAD6";
    /** Numeric keypad 7 key */
    VirtualKey[VirtualKey["VK_NUMPAD7"] = 103] = "VK_NUMPAD7";
    /** Numeric keypad 8 key */
    VirtualKey[VirtualKey["VK_NUMPAD8"] = 104] = "VK_NUMPAD8";
    /** Numeric keypad 9 key */
    VirtualKey[VirtualKey["VK_NUMPAD9"] = 105] = "VK_NUMPAD9";
    /** Multiply key */
    VirtualKey[VirtualKey["VK_MULTIPLY"] = 106] = "VK_MULTIPLY";
    /** Add key */
    VirtualKey[VirtualKey["VK_ADD"] = 107] = "VK_ADD";
    /** Separator key */
    VirtualKey[VirtualKey["VK_SEPARATOR"] = 108] = "VK_SEPARATOR";
    /** Subtract key */
    VirtualKey[VirtualKey["VK_SUBTRACT"] = 109] = "VK_SUBTRACT";
    /** Decimal key */
    VirtualKey[VirtualKey["VK_DECIMAL"] = 110] = "VK_DECIMAL";
    /** Divide key */
    VirtualKey[VirtualKey["VK_DIVIDE"] = 111] = "VK_DIVIDE";
    /** F1 key */
    VirtualKey[VirtualKey["VK_F1"] = 112] = "VK_F1";
    VirtualKey[VirtualKey["VK_F2"] = 113] = "VK_F2";
    VirtualKey[VirtualKey["VK_F3"] = 114] = "VK_F3";
    VirtualKey[VirtualKey["VK_F4"] = 115] = "VK_F4";
    VirtualKey[VirtualKey["VK_F5"] = 116] = "VK_F5";
    VirtualKey[VirtualKey["VK_F6"] = 117] = "VK_F6";
    VirtualKey[VirtualKey["VK_F7"] = 118] = "VK_F7";
    VirtualKey[VirtualKey["VK_F8"] = 119] = "VK_F8";
    VirtualKey[VirtualKey["VK_F9"] = 120] = "VK_F9";
    VirtualKey[VirtualKey["VK_F10"] = 121] = "VK_F10";
    VirtualKey[VirtualKey["VK_F11"] = 122] = "VK_F11";
    VirtualKey[VirtualKey["VK_F12"] = 123] = "VK_F12";
    VirtualKey[VirtualKey["VK_F13"] = 124] = "VK_F13";
    VirtualKey[VirtualKey["VK_F14"] = 125] = "VK_F14";
    VirtualKey[VirtualKey["VK_F15"] = 126] = "VK_F15";
    VirtualKey[VirtualKey["VK_F16"] = 127] = "VK_F16";
    VirtualKey[VirtualKey["VK_F17"] = 128] = "VK_F17";
    VirtualKey[VirtualKey["VK_F18"] = 129] = "VK_F18";
    VirtualKey[VirtualKey["VK_F19"] = 130] = "VK_F19";
    VirtualKey[VirtualKey["VK_F20"] = 131] = "VK_F20";
    VirtualKey[VirtualKey["VK_F21"] = 132] = "VK_F21";
    VirtualKey[VirtualKey["VK_F22"] = 133] = "VK_F22";
    VirtualKey[VirtualKey["VK_F23"] = 134] = "VK_F23";
    VirtualKey[VirtualKey["VK_F24"] = 135] = "VK_F24";
    /** NUM LOCK key */
    VirtualKey[VirtualKey["VK_NUMLOCK"] = 144] = "VK_NUMLOCK";
    /** SCROLL LOCK key */
    VirtualKey[VirtualKey["VK_SCROLL"] = 145] = "VK_SCROLL";
    /** Left SHIFT key */
    VirtualKey[VirtualKey["VK_LSHIFT"] = 160] = "VK_LSHIFT";
    /** Right SHIFT key */
    VirtualKey[VirtualKey["VK_RSHIFT"] = 161] = "VK_RSHIFT";
    /** Left CONTROL key */
    VirtualKey[VirtualKey["VK_LCONTROL"] = 162] = "VK_LCONTROL";
    /** Right CONTROL key */
    VirtualKey[VirtualKey["VK_RCONTROL"] = 163] = "VK_RCONTROL";
    /** Left ALT key */
    VirtualKey[VirtualKey["VK_LMENU"] = 164] = "VK_LMENU";
    /** Right ALT key */
    VirtualKey[VirtualKey["VK_RMENU"] = 165] = "VK_RMENU";
    /** Browser Back key */
    VirtualKey[VirtualKey["VK_BROWSER_BACK"] = 166] = "VK_BROWSER_BACK";
    /** Browser Forward key */
    VirtualKey[VirtualKey["VK_BROWSER_FORWARD"] = 167] = "VK_BROWSER_FORWARD";
    /** Browser Refresh key */
    VirtualKey[VirtualKey["VK_BROWSER_REFRESH"] = 168] = "VK_BROWSER_REFRESH";
    /** Browser Stop key */
    VirtualKey[VirtualKey["VK_BROWSER_STOP"] = 169] = "VK_BROWSER_STOP";
    /** Browser Search key */
    VirtualKey[VirtualKey["VK_BROWSER_SEARCH"] = 170] = "VK_BROWSER_SEARCH";
    /** Browser Favorites key */
    VirtualKey[VirtualKey["VK_BROWSER_FAVORITES"] = 171] = "VK_BROWSER_FAVORITES";
    /** Browser Start and Home key */
    VirtualKey[VirtualKey["VK_BROWSER_HOME"] = 172] = "VK_BROWSER_HOME";
    /** Volume Mute key */
    VirtualKey[VirtualKey["VK_VOLUME_MUTE"] = 173] = "VK_VOLUME_MUTE";
    /** Volume Down key */
    VirtualKey[VirtualKey["VK_VOLUME_DOWN"] = 174] = "VK_VOLUME_DOWN";
    /** Volume Up key */
    VirtualKey[VirtualKey["VK_VOLUME_UP"] = 175] = "VK_VOLUME_UP";
    /** Next Track key */
    VirtualKey[VirtualKey["VK_MEDIA_NEXT_TRACK"] = 176] = "VK_MEDIA_NEXT_TRACK";
    /** Previous Track key */
    VirtualKey[VirtualKey["VK_MEDIA_PREV_TRACK"] = 177] = "VK_MEDIA_PREV_TRACK";
    /** Stop Media key  */
    VirtualKey[VirtualKey["VK_MEDIA_STOP"] = 178] = "VK_MEDIA_STOP";
    /** Play / Pause Media key */
    VirtualKey[VirtualKey["VK_MEDIA_PLAY_PAUSE"] = 179] = "VK_MEDIA_PLAY_PAUSE";
    /** Start Mail key */
    VirtualKey[VirtualKey["VK_LAUNCH_MAIL"] = 180] = "VK_LAUNCH_MAIL";
    /** Select Media key */
    VirtualKey[VirtualKey["VK_LAUNCH_MEDIA_SELECT"] = 181] = "VK_LAUNCH_MEDIA_SELECT";
    /** Start Application 1 key */
    VirtualKey[VirtualKey["VK_LAUNCH_APP1"] = 182] = "VK_LAUNCH_APP1";
    /** Start Application 2 key */
    VirtualKey[VirtualKey["VK_LAUNCH_APP2"] = 183] = "VK_LAUNCH_APP2";
    /** Used for miscellaneous characters; it can vary by keyboard.For the US standard keyboard, the ';:' key  */
    VirtualKey[VirtualKey["VK_OEM_1"] = 186] = "VK_OEM_1";
    /** For any country / region, the '+' key */
    VirtualKey[VirtualKey["VK_OEM_PLUS"] = 187] = "VK_OEM_PLUS";
    /** For any country / region, the ',' key */
    VirtualKey[VirtualKey["VK_OEM_COMMA"] = 188] = "VK_OEM_COMMA";
    /** For any country / region, the '-' key */
    VirtualKey[VirtualKey["VK_OEM_MINUS"] = 189] = "VK_OEM_MINUS";
    /** For any country / region, the '.' key */
    VirtualKey[VirtualKey["VK_OEM_PERIOD"] = 190] = "VK_OEM_PERIOD";
    /** Used for miscellaneous characters; it can vary by keyboard.For the US standard keyboard, the '/?' key */
    VirtualKey[VirtualKey["VK_OEM_2"] = 191] = "VK_OEM_2";
    /** Used for miscellaneous characters; it can vary by keyboard.For the US standard keyboard, the '`~' key  */
    VirtualKey[VirtualKey["VK_OEM_3"] = 192] = "VK_OEM_3";
    /** Used for miscellaneous characters; it can vary by keyboard.For the US standard keyboard, the '[{' key */
    VirtualKey[VirtualKey["VK_OEM_4"] = 219] = "VK_OEM_4";
    /** Used for miscellaneous characters; it can vary by keyboard.For the US standard keyboard, the '\|' key */
    VirtualKey[VirtualKey["VK_OEM_5"] = 220] = "VK_OEM_5";
    /** Used for miscellaneous characters; it can vary by keyboard.For the US standard keyboard, the ']}' key */
    VirtualKey[VirtualKey["VK_OEM_6"] = 221] = "VK_OEM_6";
    /**
     * Used for miscellaneous characters; it can vary by keyboard.For the US standard keyboard,
     * the 'single-quote/double-quote' key
     */
    VirtualKey[VirtualKey["VK_OEM_7"] = 222] = "VK_OEM_7";
    /** Used for miscellaneous characters; it can vary by keyboard. */
    VirtualKey[VirtualKey["VK_OEM_8"] = 223] = "VK_OEM_8";
    /** The <> keys on the US standard keyboard, or the \\| key on the non - US 102 - key keyboard  */
    VirtualKey[VirtualKey["VK_OEM_102"] = 226] = "VK_OEM_102";
    /** IME PROCESS key */
    VirtualKey[VirtualKey["VK_PROCESSKEY"] = 229] = "VK_PROCESSKEY";
    /**
     * Used to pass Unicode characters as if they were keystrokes.
     * The VK_PACKET key is the low word of a 32 - bit Virtual Key value used for non - keyboard input methods.
     * For more information, see Remark in KEYBDINPUT, SendInput, WM_KEYDOWN, and WM_KEYUP
     */
    VirtualKey[VirtualKey["VK_PACKET"] = 231] = "VK_PACKET";
    /** Attn key */
    VirtualKey[VirtualKey["VK_ATTN"] = 246] = "VK_ATTN";
    /** CrSel key */
    VirtualKey[VirtualKey["VK_CRSEL"] = 247] = "VK_CRSEL";
    /** ExSel key */
    VirtualKey[VirtualKey["VK_EXSEL"] = 248] = "VK_EXSEL";
    /** Erase EOF key */
    VirtualKey[VirtualKey["VK_EREOF"] = 249] = "VK_EREOF";
    /** Play key */
    VirtualKey[VirtualKey["VK_PLAY"] = 250] = "VK_PLAY";
    /** Zoom key */
    VirtualKey[VirtualKey["VK_ZOOM"] = 251] = "VK_ZOOM";
    /** Reserved */
    VirtualKey[VirtualKey["VK_NONAME"] = 252] = "VK_NONAME";
    /** PA1 key */
    VirtualKey[VirtualKey["VK_PA1"] = 253] = "VK_PA1";
    /** Clear key */
    VirtualKey[VirtualKey["VK_OEM_CLEAR"] = 254] = "VK_OEM_CLEAR";
})(VirtualKey || (VirtualKey = {}));
