// win32-def/dist/lib/consts/winmsg.enum.js
/**
 * @link https://learn.microsoft.com/en-us/windows/win32/winmsg/windowing
 * @link https://learn.microsoft.com/zh-cn/windows/win32/winmsg/window-notifications
 */
export var WIN_MSG;
(function (WIN_MSG) {
    /** https://docs.microsoft.com/zh-cn/windows/win32/winmsg/window-messages */
    WIN_MSG[WIN_MSG["MN_GETHMENU"] = 481] = "MN_GETHMENU";
    WIN_MSG[WIN_MSG["WM_ERASEBKGND"] = 20] = "WM_ERASEBKGND";
    WIN_MSG[WIN_MSG["WM_GETFONT"] = 49] = "WM_GETFONT";
    WIN_MSG[WIN_MSG["WM_GETTEXT"] = 13] = "WM_GETTEXT";
    WIN_MSG[WIN_MSG["WM_GETTEXTLENGTH"] = 14] = "WM_GETTEXTLENGTH";
    WIN_MSG[WIN_MSG["WM_SETFONT"] = 48] = "WM_SETFONT";
    WIN_MSG[WIN_MSG["WM_SETICON"] = 128] = "WM_SETICON";
    WIN_MSG[WIN_MSG["WM_SETTEXT"] = 12] = "WM_SETTEXT";
    /** https://docs.microsoft.com/zh-cn/windows/win32/winmsg/window-notifications */
    WIN_MSG[WIN_MSG["WM_ACTIVATEAPP"] = 28] = "WM_ACTIVATEAPP";
    WIN_MSG[WIN_MSG["WM_CANCELMODE"] = 31] = "WM_CANCELMODE";
    WIN_MSG[WIN_MSG["WM_CHILDACTIVATE"] = 34] = "WM_CHILDACTIVATE";
    WIN_MSG[WIN_MSG["WM_CLOSE"] = 16] = "WM_CLOSE";
    WIN_MSG[WIN_MSG["WM_CREATE"] = 1] = "WM_CREATE";
    WIN_MSG[WIN_MSG["WM_DESTROY"] = 2] = "WM_DESTROY";
    WIN_MSG[WIN_MSG["WM_ENABLE"] = 10] = "WM_ENABLE";
    WIN_MSG[WIN_MSG["WM_ENTERSIZEMOVE"] = 561] = "WM_ENTERSIZEMOVE";
    WIN_MSG[WIN_MSG["WM_EXITSIZEMOVE"] = 562] = "WM_EXITSIZEMOVE";
    WIN_MSG[WIN_MSG["WM_GETICON"] = 127] = "WM_GETICON";
    WIN_MSG[WIN_MSG["WM_GETMINMAXINFO"] = 36] = "WM_GETMINMAXINFO";
    WIN_MSG[WIN_MSG["WM_INPUTLANGCHANGE"] = 81] = "WM_INPUTLANGCHANGE";
    WIN_MSG[WIN_MSG["WM_INPUTLANGCHANGEREQUEST"] = 80] = "WM_INPUTLANGCHANGEREQUEST";
    WIN_MSG[WIN_MSG["WM_MOVE"] = 3] = "WM_MOVE";
    WIN_MSG[WIN_MSG["WM_MOVING"] = 534] = "WM_MOVING";
    WIN_MSG[WIN_MSG["WM_NCACTIVATE"] = 134] = "WM_NCACTIVATE";
    WIN_MSG[WIN_MSG["WM_NCCALCSIZE"] = 131] = "WM_NCCALCSIZE";
    WIN_MSG[WIN_MSG["WM_NCCREATE"] = 129] = "WM_NCCREATE";
    WIN_MSG[WIN_MSG["WM_NCDESTROY"] = 130] = "WM_NCDESTROY";
    WIN_MSG[WIN_MSG["WM_NULL"] = 0] = "WM_NULL";
    WIN_MSG[WIN_MSG["WM_QUERYDRAGICON"] = 55] = "WM_QUERYDRAGICON";
    WIN_MSG[WIN_MSG["WM_QUERYOPEN"] = 19] = "WM_QUERYOPEN";
    WIN_MSG[WIN_MSG["WM_QUIT"] = 18] = "WM_QUIT";
    WIN_MSG[WIN_MSG["WM_SHOWWINDOW"] = 24] = "WM_SHOWWINDOW";
    WIN_MSG[WIN_MSG["WM_SIZE"] = 5] = "WM_SIZE";
    WIN_MSG[WIN_MSG["WM_SIZING"] = 532] = "WM_SIZING";
    WIN_MSG[WIN_MSG["WM_STYLECHANGED"] = 125] = "WM_STYLECHANGED";
    WIN_MSG[WIN_MSG["WM_STYLECHANGING"] = 124] = "WM_STYLECHANGING";
    WIN_MSG[WIN_MSG["WM_THEMECHANGED"] = 794] = "WM_THEMECHANGED";
    WIN_MSG[WIN_MSG["WM_USERCHANGED"] = 84] = "WM_USERCHANGED";
    WIN_MSG[WIN_MSG["WM_WINDOWPOSCHANGED"] = 71] = "WM_WINDOWPOSCHANGED";
    WIN_MSG[WIN_MSG["WM_WINDOWPOSCHANGING"] = 70] = "WM_WINDOWPOSCHANGING";
    /** https://learn.microsoft.com/en-us/windows/win32/dataxchg/wm-copydata */
    WIN_MSG[WIN_MSG["WM_COPYDATA"] = 74] = "WM_COPYDATA";
    // https://learn.microsoft.com/en-us/windows/win32/menurc/menu-notifications
    WIN_MSG[WIN_MSG["WM_COMMAND"] = 273] = "WM_COMMAND";
    WIN_MSG[WIN_MSG["WM_CONTEXTMENU"] = 123] = "WM_CONTEXTMENU";
    WIN_MSG[WIN_MSG["WM_ENTERMENULOOP"] = 529] = "WM_ENTERMENULOOP";
    WIN_MSG[WIN_MSG["WM_EXITMENULOOP"] = 530] = "WM_EXITMENULOOP";
    WIN_MSG[WIN_MSG["WM_GETTITLEBARINFOEX"] = 831] = "WM_GETTITLEBARINFOEX";
    WIN_MSG[WIN_MSG["WM_MENUCOMMAND"] = 294] = "WM_MENUCOMMAND";
    WIN_MSG[WIN_MSG["WM_MENUDRAG"] = 291] = "WM_MENUDRAG";
    WIN_MSG[WIN_MSG["WM_MENUGETOBJECT"] = 292] = "WM_MENUGETOBJECT";
    WIN_MSG[WIN_MSG["WM_MENURBUTTONUP"] = 290] = "WM_MENURBUTTONUP";
    WIN_MSG[WIN_MSG["WM_NEXTMENU"] = 531] = "WM_NEXTMENU";
    WIN_MSG[WIN_MSG["WM_UNINITMENUPOPUP"] = 293] = "WM_UNINITMENUPOPUP";
})(WIN_MSG || (WIN_MSG = {}));
