// win32-def/dist/lib/consts/winuser-mouseinput.enum.js
/**
 * A simulated keyboard event
 * @link https://learn.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-mouseinput
 */
export var MOUSEINPUT;
(function (MOUSEINPUT) {
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_MOVE"] = 1] = "MOUSEEVENTF_MOVE";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_LEFTDOWN"] = 2] = "MOUSEEVENTF_LEFTDOWN";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_LEFTUP"] = 4] = "MOUSEEVENTF_LEFTUP";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_RIGHTDOWN"] = 8] = "MOUSEEVENTF_RIGHTDOWN";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_RIGHTUP"] = 16] = "MOUSEEVENTF_RIGHTUP";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_MIDDLEDOWN"] = 32] = "MOUSEEVENTF_MIDDLEDOWN";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_MIDDLEUP"] = 64] = "MOUSEEVENTF_MIDDLEUP";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_XDOWN"] = 128] = "MOUSEEVENTF_XDOWN";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_XUP"] = 256] = "MOUSEEVENTF_XUP";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_WHEEL"] = 2048] = "MOUSEEVENTF_WHEEL";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_HWHEEL"] = 4096] = "MOUSEEVENTF_HWHEEL";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_MOVE_NOCOALESCE"] = 8192] = "MOUSEEVENTF_MOVE_NOCOALESCE";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_VIRTUALDESK"] = 16384] = "MOUSEEVENTF_VIRTUALDESK";
    MOUSEINPUT[MOUSEINPUT["MOUSEEVENTF_ABSOLUTE"] = 32768] = "MOUSEEVENTF_ABSOLUTE";
})(MOUSEINPUT || (MOUSEINPUT = {}));
